### DISCUSSION TITLE
Read xmltv with PHP

### DISCUSSION THREAD

**Discussion by LordHelmchen666** [Moonbase59/azuracast_xmltv#4] [Show and tell]
I have done a quick test to show what is possible. Here is a quick example
```
<html>
<head>
</head>
<body>

<?php
$path = 'https://YOURURL/xmltv/XMLTV-FILE.xml';

    $xml=simplexml_load_file($path);
    #print_r($xml);
            $me = "sub-title";
            foreach($xml->programme as $item){
                echo "<b>" .date("G:i", strtotime(substr($item["start"], 0,  -6))) . " Uhr </b>" . date("d.m.Y", strtotime(substr($item["start"], 0,  -6))) . '<br>';
                echo " bis  " .date("G:i d.m.Y", strtotime(substr($item["stop"], 0,  -6))) . ' <br>';
                echo "<b><i> ".$item->title. "</i></b><br>";
                echo "<i>".$item->$me. "</i><br>";
                echo "".nl2br($item->desc). "<br>";
                echo "<br><hr>";
            }

?> 


</body>
</html>

```
https://demo.helmut-mueller-online.de/readxmltv.php

**Comment by Moonbase59** [comment]
Cool, I love creative ideas!

Now we only have to convince @BusterNeece that playlist and schedule item descriptions would be _really useful_.  Just imagine getting nicer show descriptions than "Playlist: MyPlayListName". ;-)

I think using PHP _SimpleXML_ is a good choice, but be warned that it becomes a real memory hog if you start parsing larger EPG files, which can happen when using AzuraCast servers with many stations, or _xTeve_. My _xTeVe_ `xteve.xml` is already ~20 MB in size, and I’ve seen some using ~50 MB. (_xTeVe_ aggregates all sources you give it and produces two large files containing all stations and their EPG. Another reason to use the gzip format…)

For _creating_ an XMLTV EPG in PHP I’d use the well-written and actively developed XMLTV library by @jaylinski: https://github.com/jaylinski/xmltv.

**Reply by BusterNeece** [reply]
I mean, we can do that...did you submit it?

**Reply by Moonbase59** [reply]
@BusterNeece: I _think_ I put something like that on the FR page once or twice… Only that page is so limited, and I wanted to see a real use case first, apart from displaying that on the schedules page as a hover text or the like (might still be nice). We really have to get this Discord thingy going…

I still wonder what the more logical place(s) would be: Put a, say 256-512 char description in _schedules_, _playlists_ (description already there), or _both_?

I think this should be a two-step process:
- Allow entering a playlist description (allowing newlines) for a _playlist_.
- When creating a _scheduled item_, also allow entering a description (and add the field to DB+API), which can be prefilled from the playlist description. (A scheduled item could also come from a live DJ show or remote stream (syndicated content), and to have a description for _that_ would _really_ be neat.)
- Allow a one-liner as kind of a "heading" (title of the description), which would make the EPG just beautiful, since _that_ could be used as _subtitle_ in the EPG.

Reason for not just using playlist description: Accessing the playlists needs an API key (good so), but schedules are _public_. So if the _schedules_ would have a _(sub-)title_ and _description_, we could generate a real neat EPG programme entry from other stations, too.

I’d then use the schedule’s JSON like
- `name` → EPG Programme _Title_
- `title` → EPG Programme _Subtitle_
- `description` → EPG Programme _Description_

This should guarantee the best user experience end-to-end, with the least changes, since you already thought of having a name, a title, and a description!

![Auswahl_289](https://github.com/Moonbase59/azuracast_xmltv/assets/3706922/9b4fde83-61cc-4214-af1f-37f5c7e9152d)

![Auswahl_292](https://github.com/Moonbase59/azuracast_xmltv/assets/3706922/68216dcd-8df6-43d8-b8ea-67249e08cfc3)

![Auswahl_291](https://github.com/Moonbase59/azuracast_xmltv/assets/3706922/14e8f480-c879-45ff-8533-eac0178ce36e)

---

Looking into the bright future, we might also, when all this is finished, think about adding a `composer require jaylinski/xmltv` to AzuraCast, re-emulate all my crude logic into PHP, and do what `azuracast_xmltv` does _natively_ within AzuraCast… :wink:

**Reply by Moonbase59** [reply]
EPGs support having an _icon_ per programme, we might even think about adding an _image_ per playlist/schedule. Or would that be a little overkill?

When devising these ideas, I also think of AzuraCast’s built-in public "Schedule" page—image, title, and description could well be used for a hover- or click-type "show description" an thus enrich the native software.

**Reply by Moonbase59** [reply]
Added **[AzuraCast Feature Request](https://features.azuracast.com/suggestions/504596/add-epgcompatible-image-title-description-fields-to-playlists-and-schedules-for-)**.
